Tensile/Source/lib/source/ContractionSolution.cpp

```diff
@@ -308,16 +308,31 @@ namespace Tensile
 
         if(sizeMapping.persistentKernel != 0)
         {
-            size_t persistentGroups = dynamic_cast<AMDGPU const&>(hardware).computeUnitCount
-                                      * sizeMapping.persistentKernel;
+            AMDGPU const* pAMDGPU = dynamic_cast<AMDGPU const*>(&hardware);
+            assert(pAMDGPU != nullptr);
+
+            size_t cuCount       = pAMDGPU->computeUnitCount;
+            size_t finalPKValue  = sizeMapping.persistentKernel;
             size_t problemGroups = rv.numWorkGroups.x * rv.numWorkGroups.y;
             if(sizeMapping.persistentKernelAlongBatch)
             {
                 problemGroups *= rv.numWorkGroups.z;
                 rv.numWorkGroups.z = 1;
             }
-            rv.numWorkGroups.x = std::min(persistentGroups, problemGroups);
-            rv.numWorkGroups.y = 1;
+
+            if(finalPKValue == -1)
+            {
+                // 1. Get the largest pk value (ex.3)
+                //    which can make the PK.G (ex.3*120=360) <= problemGroups (ex.433)
+                // 2. Scale by 5/8 (can try 0.5~1, to control the tiles-per-workgroup = 1~2)
+                finalPKValue = 5 * (problemGroups / cuCount) / 8;
```

Comment:
Small risk, but consider an assertion for cuCount != 0

Reply:
Done !